refactor params

deanlee opened the pull request:
1. remove functions write_param_float, read_param, read_db_bool, read_db_bytes
2. add template functions get(),put()
3. rewrite read_db_value to to eliminate memory copy

deanlee committed:
refactor params

deanlee committed:
get() return std::optional

deanlee committed:
reduce 1 line

adeebshihadeh commented:
Nice, we've been planning to do this since the initial params refactor. We also wanted to make the python and C API the same. Can we make the C API match the python one?

deanlee committed:
same api with python version

deanlee committed:
continue

deanlee committed:
refactor read_all

deanlee committed:
function lock_path&key_path

deanlee committed:
combile ifdef

deanlee committed:
refactor params::put&params::delete_value

deanlee committed:
more

deanlee committed:
rebase master

std::string_view

const params_path

faster ensure_dir_exists

Reduce memory allocation

const return

use unique_fd

inline functions

cleanup includes

add todo

static read_value

read_file_string

rebase master

ready for review

remove check

help function:getBool

rebase master & add function util::file_exists

remove duplate #include

more readable

rebase master

rebase master

adeebshihadeh commented on selfdrive/common/params.h:
```diff
@@ -1,46 +1,45 @@
-  std::string get(std::string key, bool block=false);
+  Params(std::string_view path);
+  
+  bool delete_value(std::string key);
```
let’s make this just delete to match the python

deanlee replied:
delete is a reserved C++ keyword.we can't use it as function name.

adeebshihadeh commented on selfdrive/loggerd/loggerd.cc:
```diff
@@ -514,28 +514,20 @@ kj::Array<capnp::word> gen_init_data() {
-  if (git_commit.size() > 0) {
-    init.setGitCommit(capnp::Text::Reader(git_commit.data(), git_commit.size()));
+  Params params = Params();
+  if (std::string v = params.get("GitCommit"); v.size() > 0) {
```
do we still need this check?

adeebshihadeh commented on selfdrive/loggerd/loggerd.cc:
```diff
@@ -613,7 +605,7 @@ int main(int argc, char** argv) {
   bool record_front = true;
 #ifndef QCOM2
-  record_front = Params().read_db_bool("RecordFront");
+  record_front = Params().get<bool>("RecordFront").value_or(false);
```
bool params are pretty common. maybe make a helper?

deanlee replied:
how about add a new function getBool()?

deanlee commented:
don't know why the test always failed

deanlee commented:
Need to rearrange the usage of string_view

deanlee committed:
remove std::string_view

deanlee committed:
continue

adeebshihadeh commented:
It looks like the tests are timing out in CI, probably due to bug that's causing params to block in the refactor.

deanlee committed:
fix bug in ensure_symlink

deanlee commented:
> It looks like the tests are timing out in CI, probably due to a bug that's causing params to block in the refactor.

u r right,there are bugs in ensure_symlink,which always return false

deanlee committed:
remove test_params.c

deanlee commented:
Delete obsolete test_params.c

deanlee committed:
replace static with anonymouse namespace

deanlee committed:
continue

adeebshihadeh commented on selfdrive/boardd/boardd.cc:
```diff
@@ -73,12 +73,10 @@ void safety_setter_thread() {
 
-    std::vector<char> value_vin = Params().read_db_bytes("CarVin");
-    if (value_vin.size() > 0) {
+    if (std::string vin = Params().get("CarVin"); vin.size() > 0) {
```
use two lines here

adeebshihadeh commented on selfdrive/ui/qt/offroad/settings.cc:
```diff
@@ -169,16 +168,15 @@ QWidget * developer_panel() {
 
-  std::string os_version = util::read_file("/VERSION");
-  if (os_version.size()) {
+  if (std::string os_version = util::read_file("/VERSION"); os_version.size()) {
```
use two lines

adeebshihadeh replied:
use two lines here

adeebshihadeh commented on selfdrive/loggerd/loggerd.cc:
```diff
@@ -469,28 +460,14 @@ kj::Array<capnp::word> gen_init_data() {
+  init.setGitRemote(params.get("GitRemote"));
+  init.setPassive(params.getBool("Passive"));
+  // log params
+  if (std::map<std::string, std::string> params_map; params.read_all(params_map)) {
```
use two lines here

adeebshihadeh commented on selfdrive/ui/qt/offroad/onboarding.cc:
```diff
@@ -35,7 +35,7 @@ QWidget * OnboardingWindow::terms_screen() {
   main_layout->addWidget(accept_btn, 2, 1);
   QObject::connect(accept_btn, &QPushButton::released, [=]() {
-    Params().write_db_value("HasAcceptedTerms", LATEST_TERMS_VERSION);
+    Params().put("HasAcceptedTerms", LATEST_TERMS_VERSION, strlen(LATEST_TERMS_VERSION));
```
why does this need strlen? it should be passed in as a string

adeebshihadeh commented on selfdrive/common/params.h:
```diff
@@ -1,46 +1,47 @@
+    }
+    return std::nullopt;
+  }
+  inline bool getBool(const char *param_name, bool block = false) {
```
the bool helper should take a string, otherwise we'll have `.c_str()` everywhere

adeebshihadeh commented on selfdrive/common/params.h:
```diff
@@ -1,46 +1,47 @@
+    return read_value(key.c_str(), block).value_or("");
+  }
+  template <class T>
+  std::optional<T> get(const char *param_name, bool block = false) {
```
why do we have functions defined in the header?

deanlee replied:
we use it to read other types such as float,double. for example:
`float brightness_b = params.get<float>("BRIGHTNESS_B").value_or(-1.)`

deanlee committed:
double line

deanlee committed:
getBool(std::string, bool)

deanlee committed:
better put : use if constexpr

deanlee committed:
cleanup

deanlee committed:
use std::string for param_name

deanlee committed:
rebase master,fix build error

deanlee committed:
rebase master

adeebshihadeh commented:
Can you rebase?

deanlee committed:
rebase master

rebase master

deanlee commented:
temporarily convert to draft due to a compile error

deanlee committed:
simplify

deanlee committed:
fix

deanlee committed:
cleanup

deanlee committed:
const std::string&

deanlee committed:
cleanup

deanlee commented:
ready for review

deanlee committed:
rebase master

deanlee committed:
rebase master

deanlee committed:
rebase master

deanlee committed:
rebase master

adeebshihadeh commented:
Can you rebase?

deanlee committed:
rebase master

deanlee committed:
static env_home

deanlee commented:
Converted to draft to continue adjust, the master has changed too much

deanlee committed:
send std::string to file_exist

deanlee commented:
reddy for review

deanlee committed:
rebase master

deanlee committed:
rebase master

deanlee committed:
rebase master

deanlee committed:
rebase master

deanlee committed:
fix build err

deanlee commented:
Need an interface to read vector.

deanlee committed:
continue

deanlee committed:
rebase master

pd0wm commented on selfdrive/boardd/boardd.cc:
```diff
@@ -146,13 +145,13 @@ bool usb_connect() {
   // get panda serial
-  if (auto serial = tmp_panda->get_serial(); serial) {
-    params.write_db_value("PandaDongleId", serial->c_str(), serial->length());
+  if (auto serial = panda->get_serial(); serial) {
```
```suggestion
  if (auto serial = tmp_panda->get_serial(); serial) {
```

Mistake from rebasing?

deanlee committed:
Update selfdrive/boardd/boardd.cc

Co-authored-by: Willem Melching <willem.melching@gmail.com>

pd0wm commented:
Like the changes using optional and the templates.

I'd prefer to not touch the implementations or only make minimal changes in `write_db_value`,`delete_db_value` and `read_all` (using the get method is fine though instead of reimplementing the read) unless something is actually broken. The current code might not be the shortest, but it's well commented and reasonable clear.

deanlee commented:
OIC

deanlee commented:
How about this one? replaced with a very python-like interface, and use  `util:read_file `to read file : https://github.com/commaai/openpilot/pull/20506 
It is difficult to continue to split into smaller PRs.

pd0wm commented:
That looks manageable. Thanks!
